### PR TITLE
feature/plan-115-web-copy-game-link

### PR DESCRIPTION
**Pull request by RutaB98**

### CODE REVIEW

File: E2E/pageObjects/game.page.ts [hunk 1]
```diff
@@ -31,6 +32,14 @@ export = {
     I.dontSeeElement(locator.chessPieces.container);
     I.seeElement(locator.buttons.copyLink);
     I.seeElement(locator.buttons.restartGame);
-  }
+  },
 
+  checkCopyLinkButton: () => {
+    I.waitForElement(locator.buttons.copyLink, 10);
+    I.click(locator.buttons.copyLink);
+    I.waitForElement(locator.text.linkCopiedToClipboard, 7);
```
Comment (jvalatkevicius) [inline E2E/pageObjects/game.page.ts:40]: Why are you using different explicit timeouts here? (10 and 7). I would expect default codecept timeouts to be enough.

File: E2E/tests/copyLink.test.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+Feature('Copy link');
```
Comment (jvalatkevicius) [inline E2E/tests/copyLink.test.ts:1]: This seems to be missing test that would check if the link was actually copied into the cliipboard.

File: E2E/pageObjects/game.page.ts [hunk 3]
```diff
@@ -8,6 +8,7 @@ const locator = {
   },
   text: {
     username: '#username',
+    linkCopiedToClipboard: '//*[contains(text(), "Link copied to clipboard")]',
```
Comment (jvalatkevicius) [inline E2E/pageObjects/game.page.ts:11]: Would it work, if you removed `contains` function and leave only exact match by text?

File: E2E/pageObjects/game.page.ts [hunk 4]
```diff
@@ -16,7 +17,7 @@ const locator = {
     container: '$chess-pieces-container',
   },
   buttons: {
-    copyLink: '//button/p[contains(text(), "Copy Link")]',
+    copyLink: '//*[contains(text(), "Copy Link")]',
```
Comment (jvalatkevicius) [inline E2E/pageObjects/game.page.ts:20]: Same as above

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: E2E/package-lock.json [hunk 5]
```diff
@@ -16,10 +16,6 @@
       "devDependencies": {
         "@types/node": "^12.0.0",
         "dotenv": "^10.0.0",
-        "eslint-config-prettier": "^8.3.0",
```
Comment (jvalatkevicius) [inline E2E/package-lock.json:19]: Why these were removed?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: E2E/pageObjects/game.page.ts (hunk 4 above)
Comment (KristinaKrev) [inline E2E/pageObjects/game.page.ts:20]: There are two elements with "Copy link" text on page, doesn't this find both of them?

File: E2E/pageObjects/game.page.ts [hunk 6]
```diff
@@ -31,6 +32,14 @@ export = {
...
     I.seeElement(locator.buttons.copyLink);
     I.seeElement(locator.buttons.restartGame);
-  }
+  },
 
+  checkCopyLinkButton: () => {
+    I.waitForElement(locator.buttons.copyLink, 10);
+    I.click(locator.buttons.copyLink);
+    I.waitForElement(locator.text.linkCopiedToClipboard, 7);
+    I.waitForInvisible(locator.text.linkCopiedToClipboard, 7);
```
Comment (KristinaKrev) [inline E2E/pageObjects/game.page.ts:41]: Based on Acceptance criteria, tooltip should disappear after 2 sek, is this timeout correct?

File: E2E/assertions/copyLinkAssert.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import assert = require("assert");
+import { displayPartsToString } from "typescript";
+import gamePage = require("../pageObjects/game.page");
+const { I } = inject();
+
+export = {
+    async checkIfTheLinkWasCopied () {
+      let currentUrl = await I.grabCurrentUrl();
```
Comment (jvalatkevicius) [inline E2E/assertions/copyLinkAssert.ts:8]: why `let`?

File: E2E/assertions/copyLinkAssert.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+      I.waitForElement(gamePage.locator.buttons.copyLink);
+      I.click(gamePage.locator.buttons.copyLink); 
+      I.usePlaywrightTo('check clipboard', async ({ page, browserContext }) => {
+        await browserContext.grantPermissions(['clipboard-write', 'clipboard-read']);
+        let clipboardText = await page.evaluate("navigator.clipboard.readText()");
```
Comment (jvalatkevicius) [inline E2E/assertions/copyLinkAssert.ts:13]: why `let`?

File: E2E/assertions/copyLinkAssert.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import assert = require("assert");
+import { displayPartsToString } from "typescript";
+import gamePage = require("../pageObjects/game.page");
+const { I } = inject();
+
+export = {
+    async checkIfTheLinkWasCopied () {
```
Comment (jvalatkevicius) [inline E2E/assertions/copyLinkAssert.ts:7]: Why was this moved into a separate file? Cannot we do this exact thing in the test itself?

File: E2E/assertions/copyLinkAssert.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+      let currentUrl = await I.grabCurrentUrl();
+      I.waitForElement(gamePage.locator.buttons.copyLink);
+      I.click(gamePage.locator.buttons.copyLink); 
+      I.usePlaywrightTo('check clipboard', async ({ page, browserContext }) => {
+        await browserContext.grantPermissions(['clipboard-write', 'clipboard-read']);
```
Comment (jvalatkevicius) [inline E2E/assertions/copyLinkAssert.ts:12]: is the `clipboard-write` permission needed here?

Comment (jvalatkevicius) [reply E2E/assertions/copyLinkAssert.ts:12]: Are the permissions needed in the first place?

File: E2E/pageObjects/game.page.ts [hunk 11]
```diff
@@ -16,7 +17,7 @@ const locator = {
     container: '$chess-pieces-container',
   },
   buttons: {
-    copyLink: '//button/p[contains(text(), "Copy Link")]',
+    copyLink: locate('//*[text() = "Copy Link"]').at(1),
```
Comment (jvalatkevicius) [inline E2E/pageObjects/game.page.ts:20]: What does `at(1)` do?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

**Review by jvalatkevicius — APPROVED** [review]

**Review by KristinaKrev — APPROVED** [review]

**Pull request merged**